### PR TITLE
[CARBONDATA-3555] Make move filter related methods under DataMapFilter

### PR DESCRIPTION
**Pull request by kunal642**
1. This PR will make DataMapFilter as a filter holder for the filter expression and the FilterResolver objects so that all the major API's can accept dataMapFilter as an argument.
2. Moved all the filter resolving methods inside DataMapFilter for ease of use

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/684/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/687/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/690/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/694/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/699/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/703/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/703/

**Comment by kunal642** [comment]
@ravipesala @kumarvishal09 Please review

**Comment by jackylk** [comment]
@kunal642 please describe the motivation for this PR

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/770/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/776/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/778/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/772/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/778/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/780/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/774/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/780/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/782/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/776/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/782/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/784/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/824/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/821/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/827/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/825/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/828/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/831/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/841/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/839/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/834/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/842/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/840/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/837/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/843/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/845/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/843/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/849/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/851/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/844/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/850/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/853/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/851/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/859/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/857/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/860/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/865/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/865/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/864/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/870/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/872/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonAddLoadCommand.scala [hunk 1]
```diff
@@ -180,11 +181,12 @@ case class CarbonAddLoadCommand(
 
     operationContext.setProperty(carbonTable.getTableUniqueName + "_Segment",
       model.getSegmentId)
-    val loadTablePreStatusUpdateEvent: LoadTablePreStatusUpdateEvent =
-      new LoadTablePreStatusUpdateEvent(
-        carbonTable.getCarbonTableIdentifier,
-        model)
-    OperationListenerBus.getInstance().fireEvent(loadTablePreStatusUpdateEvent, operationContext)
+    //    val loadTablePreStatusUpdateEvent: LoadTablePreStatusUpdateEvent =
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonAddLoadCommand.scala]: remove this commented code

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java [hunk 2]
```diff
@@ -72,6 +85,17 @@ public Expression getNewCopyOfExpression() {
     }
   }
 
+  public DataMapFilter(SegmentProperties properties, CarbonTable table, Expression expression) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java]: Move this constructor above getNewCopyOfExpression method

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java [hunk 3]
```diff
@@ -72,6 +85,17 @@ public Expression getNewCopyOfExpression() {
...
+    this.expression = expression;
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java]: call above constructor this(expression, table, false)

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java [hunk 4]
```diff
@@ -115,10 +139,12 @@ public DataMapFilter(FilterResolverIntf resolver) {
     this.resolver = resolver;
   }
 
-  private void resolve() {
+  public void resolve(boolean lazyResolve) {
     if (expression != null) {
-      table.processFilterExpression(expression, null, null);
-      resolver = CarbonTable.resolveFilter(expression, table.getAbsoluteTableIdentifier());
+      processFilterExpression();
+      if (!lazyResolve) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java]: please add comment when lazyResolve will be true and false

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java [hunk 5]
```diff
@@ -174,15 +174,18 @@
             readCommittedScope);
 
     // process and resolve the expression
-    Expression filter = getFilterPredicates(job.getConfiguration());
+    DataMapFilter dataMapFilter = getFilterPredicates(job.getConfiguration());
     // this will be null in case of corrupt schema file.
     PartitionInfo partitionInfo = carbonTable.getPartitionInfo(carbonTable.getTableName());
 
+    if (dataMapFilter != null) {
+      dataMapFilter.resolve(false);
```
Comment (kumarvishal09) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: In data map filter class add a method which will resolve and process

**Comment by kumarvishal09** [comment]
@kunal642 Please revert back formatting changes

File: core/src/main/java/org/apache/carbondata/core/scan/executor/impl/AbstractQueryExecutor.java [hunk 6]
```diff
@@ -329,18 +329,12 @@ private void updateColumns(QueryModel queryModel, List<ColumnSchema> columnsInTa
...
-          properties.getDimensions(),
-          properties.getMeasures(),
-          new ArrayList<CarbonDimension>());
-      QueryModel.processFilterExpression(processVO, expression, null, null, queryModel.getTable());
-      // Optimize Filter Expression and fit RANGE filters is conditions apply.
-      FilterOptimizer rangeFilterOptimizer = new RangeFilterOptmizer(expression);
-      rangeFilterOptimizer.optimizeFilter();
-      queryModel.setFilterExpressionResolverTree(
-          CarbonTable.resolveFilter(expression, queryModel.getAbsoluteTableIdentifier()));
+    if (queryModel.getDataMapFilter() != null) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/scan/executor/impl/AbstractQueryExecutor.java]: Can u please revisit this code

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonAddLoadCommand.scala (hunk 1 above)
Comment (kunal642) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonAddLoadCommand.scala]: done

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java (hunk 2 above)
Comment (kunal642) [reply core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java]: done

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java (hunk 3 above)
Comment (kunal642) [reply core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java]: done

File: core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java (hunk 4 above)
Comment (kunal642) [reply core/src/main/java/org/apache/carbondata/core/datamap/DataMapFilter.java]: added in resolve method

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java (hunk 5 above)
Comment (kunal642) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: done

File: core/src/main/java/org/apache/carbondata/core/scan/executor/impl/AbstractQueryExecutor.java (hunk 6 above)
Comment (kunal642) [reply core/src/main/java/org/apache/carbondata/core/scan/executor/impl/AbstractQueryExecutor.java]: done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/884/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/890/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/892/

**Comment by kunal642** [comment]
@kumarvishal09 Handled the comments

**Comment by kumarvishal09** [comment]
@kunal642 Please revert formatting changes

**Comment by kumarvishal09** [comment]
LGTM, @kunal642 Please revert the formatting changes

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/924/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/930/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/932/

**Comment by kunal642** [comment]
@kumarvishal09 added a new commit for a fix in datasource module..Please check again

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/946/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/937/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/947/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/939/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/948/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/949/

**Comment by kumarvishal09** [comment]
LGTM

**Pull request closed without merging**